consensus/common/src/main/java/org/hyperledger/besu/consensus/common/MigratingMiningCoordinator.java

```diff
@@ -127,17 +128,27 @@ public void onBlockAdded(final BlockAddedEvent event) {
     final long currentBlock = event.getBlock().getHeader().getNumber();
     final MiningCoordinator nextMiningCoordinator =
         miningCoordinatorSchedule.getFork(currentBlock + 1).getValue();
+
     if (activeMiningCoordinator != nextMiningCoordinator) {
       LOG.trace(
           "Migrating mining coordinator at block {} from {} to {}",
           currentBlock,
           activeMiningCoordinator.getClass().getSimpleName(),
           nextMiningCoordinator.getClass().getSimpleName());
-      activeMiningCoordinator.stop();
-      activeMiningCoordinator = nextMiningCoordinator;
-      startActiveMiningCoordinator();
-    }
-    if (activeMiningCoordinator instanceof BlockAddedObserver) {
+
+      final Runnable stopActiveCoordinatorTask = () -> activeMiningCoordinator.stop();
+      final Runnable startNextCoordinatorTask =
+          () -> {
+            activeMiningCoordinator = nextMiningCoordinator;
+            startActiveMiningCoordinator();
+            if (activeMiningCoordinator instanceof BlockAddedObserver) {
+              ((BlockAddedObserver) activeMiningCoordinator).onBlockAdded(event);
+            }
+          };
+
+      CompletableFuture.runAsync(stopActiveCoordinatorTask).thenRun(startNextCoordinatorTask);
```

Comment:
What happens if the application tries to get hold of the mining coordinator in between `stopActiveCoordinatorTask` and `startNextCoordinatorTask`? We will end up giving them a mining coordinator instance that is in the process of being stopped.

The `stop()` implementation on `AbstractMiningCoordinator` is synchronized, I wonder if running as an async callback might introduce some side effects that we haven't thought of.

Reply:
Although, it looks like we synchronize it just for the sake of keeping the MiningCoordinator internal state consistent and to avoid concurrent `shutdown()` calls to the executor, etc. So it should be safe 🤷

Reply:
> What happens if the application tries to get hold of the mining coordinator in between stopActiveCoordinatorTask and startNextCoordinatorTask? 

We delegate to BftMiningCoordinator.stop() which handles graceful shutdown:
https://github.com/hyperledger/besu/blob/93a599abf272f3214911df15d2b11b80b8bf3014/consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/blockcreation/BftMiningCoordinator.java#L86-L100

It waits for the bftProcessor to stop before shutting the executor down.
Not sure if that's what you mean by the application getting hold of mining coordinator?

BftProcessor is run on a single threaded executor so shouldn't need to synchronise (although that's what's causing the unintended consequences I'm attempting to fix). I think AbstractMiningCoordinator tasks are async but would need to double check.

Reply:
By the way, AbstractMiningCoordinator isn't used by the BFT implementations.

Reply:
Think is fine. It's possible that the some the MiningCoordinator methods could be called whilst in the process of one shutting down but couldn't see any problems with that. Some messages will probably be dropped but as soon as we start this we are done before the original consensus mechanism so think it's ok.